mobile_verifier/src/heartbeats/last_location.rs

```diff
@@ -56,9 +80,37 @@ impl LocationCache {
         }
     }
 
-    async fn fetch_from_db_and_set(
+    pub async fn set(&self, hotspot: &PublicKeyBinary, last_location: LastLocation) {
+        self.maybe_cache_last_location(hotspot, Some(last_location))
+            .await;
+    }
+
+    pub async fn get(
+        &self,
+        hotspot: &PublicKeyBinary,
+        heartbeat_timestamp: DateTime<Utc>,
+    ) -> anyhow::Result<Option<LastLocation>> {
+        let location = match self.locations.get(hotspot).await {
+            Some(last_location) => {
+                // The value may still be cached according to the system clock
+                // but not valid based on the time of the heartbeat in question.
+                let last = *last_location;
+                last.filter(|l| l.still_valid(heartbeat_timestamp))
```

Comment:
What is the scenario that makes this check necessary?

Reply:
I went back and forth on this for a while. 
The short answer is testing.

This is the only way I could think of to ensure we're keeping the contract of 24 hours from a given heartbeat. 

Time is not mockable in `retainer::Cache`. Using different cache durations didn't seem like something I wanted to do. And writing a test that waited significant amounts of real time is a no go.

Given @kurotych's question about caching None values at all, I am wondering if we should fallback to the database in a different way.

Reply:
The cache duration is derived from the value being cached. 

We could make the cache bust during a test, but that still requires waiting some amount of time during the test.

The more I think about it, the cache is a convenience, but the actual contract of the values being returned is that they are not more than 24 hours from the time of a given heartbeat.

The query achieves that, this filter line achieves that. 

We will also be rethinking how these values are applicable when we figure out how to process heartbeats in a non-CBRS world.